synapse/cores/common.py

```diff
@@ -2619,6 +2622,109 @@ def _popCoreXact(self):
     def _getCoreXact(self, size):
         raise NoSuchImpl(name='_getCoreXact')
 
+    def _onAddSynTagDark(self, mesg):
+        tufo = mesg[1].get('tufo')  # type: tuple(str, dict)
+        tag = mesg[1].get('tag')  # type: str
+        self.addTufoDark(tufo, 'tag', tag)
+
+    def _onDelSynTagDark(self, mesg):
+        tufo = mesg[1].get('tufo')  #type: ((str, dict))
+        tag = mesg[1].get('tag')
+        self.delTufoDark(tufo, 'tag', tag)
+
+    def addTufoDark(self, tufo, darktype, name):
+        '''
+        Add the tufo to a dark type with a given name.
+
+        Dark types get their own index and can be used to quickly pull sets of
+        rows.  While similar to Dsets, these are intended for implementing
+        features inside of Synapse directly.
+
+        Args:
+            tufo ((str, dict)): Tufo to add to the darktype.
+            darktype (str): Dark type to add.
+            name (str): Named type value.
+
+        Returns:
+            None: Returns None.
+        '''
+        dark = tufo[0][::-1]
+        if self.getRowsByIdProp(dark, '_:dark:%s' % darktype, valu=name):
+            return
+
+        rows = [(dark, '_:dark:%s' % darktype, name, now())]
+        self.addRows(rows)
+        # NOTE: may add this if a use cases occurs
+        # self.fire('syn:dark:add', darktype=darktype, name=name, tufo=tufo, dark=dark)
+        self.fire('syn:dark:add:%s' % darktype, dark=dark, name=name, tufo=tufo)
+
+    def delTufoDark(self, tufo, darktype, name):
```

Comment:
"darktype" should really be "name" or and "name" should really be "valu" ( especially since it may ultimately be integers for some dark rows / props )